src/lib/mina_stdlib/hashtblh.ml

```diff
@@ -0,0 +1,35 @@
+module Make (Key : sig
+  type _ t
+
+  val compare : _ t -> _ t -> int
+
+  val hash : _ t -> int
+
+  val sexp_of_t : _ t -> Sexp.t
+end) =
+struct
+  type e = E : 'a -> e
+
+  let unsafe_cast (E x : e) : 'a = Obj.magic x
```

Comment:
Obj.magic is the evilest evil. I would refuse most PR using it 😁

Reply:
This is dependent type, and you can't encode this in type safe manner without it.